File: LeetCodeCPP/933.number-of-recent-calls.cpp
/*
 * @lc app=leetcode id=933 lang=cpp
 *
 * [933] Number of Recent Calls
 */

#include <vector>
#include <iostream>
using namespace std;

// @lc code=start
class RecentCounter
{
private:
    vector<int> requests;

public:
    RecentCounter()
    {
    }

    int ping(int t)
    {
        requests.push_back(t);

        while (requests.front() < t - 3000)
        {
            requests.erase(requests.begin());
        }

        return requests.size();
    }
};

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */
// @lc code=end

int main()
{
    RecentCounter counter;

    cout << "ping: " << counter.ping(642) << endl;
    cout << "ping: " << counter.ping(1849) << endl;
    cout << "ping: " << counter.ping(4921) << endl;
    cout << "ping: " << counter.ping(5936) << endl;
    cout << "ping: " << counter.ping(5957) << endl;
}

File: LeetCodePython/933.number-of-recent-calls.py
#
# @lc app=leetcode id=933 lang=python3
#
# [933] Number of Recent Calls
#
from collections import deque


# @lc code=start
class RecentCounter:
    def __init__(self):
        self._counter = 0
        self._request_queue = deque()

    def ping(self, t: int) -> int:
        self._request_queue.append(t)
        self._counter += 1

        while self._request_queue[0] < t - 3000:
            self._request_queue.popleft()
            self._counter -= 1

        return self._counter


# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)
# @lc code=end
